collectors/modality-probe-debug-collector/src/lib.rs

```diff
@@ -0,0 +1,1169 @@
+        self.mem_accessor
+            .borrow_mut()
+            .read_32(self.overwrite_seqn_low_addr)
+    }
+
+    fn snap_storage(&self, index: usize) -> Result<LogEntry, DebugCollectorError> {
+        let raw: u32 = self
+            .mem_accessor
+            .borrow_mut()
+            .read_32(self.storage_addr + (size_of::<LogEntry>() * index) as u64)?;
+        // Safe because entry is already in memory as a valid LogEntry
+        Ok(unsafe { LogEntry::new_unchecked(raw) })
+    }
+}
+
+/// Used to write to probe's "overwrite_priority" field
+struct PriorityWriter {
+    /// Memory accessor used to write to device memoryt
+    mem_accessor: Rc<RefCell<dyn MemoryAccessor>>,
+    /// Address of priority field
+    priority_field_addr: Word,
+}
+
+impl PriorityWriter {
+    fn write(&mut self, level: u32) -> Result<(), DebugCollectorError> {
+        self.mem_accessor
+            .borrow_mut()
+            .write_32(self.priority_field_addr, level)
+    }
+}
+
+/// Log collector for a single probe
+pub struct Collector {
+    /// Sequence number of next report
+    seq_num: u64,
+    /// Reader used to read the probe's FencedRingBuffer
+    reader: FencedReader<LogEntry, MemorySnapper>,
+    /// Allocated buffer for reading the log into
+    rbuf: Vec<WholeEntry<LogEntry>>,
+    /// Processed clocks backing storage
+    clocks: Vec<LogicalClock>,
+    /// Used to write to the probe's "overwrite_priority" field
+    priority_writer: PriorityWriter,
+}
+
+impl Collector {
+    /// Initialize collector by reading probe information
+    pub fn initialize(
+        probe_addr: &ProbeAddr,
+        mem_accessor: Rc<RefCell<dyn MemoryAccessor>>,
+    ) -> Result<Self, DebugCollectorError> {
+        let addr = match *probe_addr {
+            ProbeAddr::Addr(addr) => addr,
+            // Read storage address from pointer
+            ProbeAddr::PtrAddr(addr) => mem_accessor.borrow_mut().read_word(addr)?,
+        };
+        let mut mem_borrow = mem_accessor.borrow_mut();
+        // Get address of DynamicHistory
+        let hist_addr = mem_borrow.read_word(addr + history_ptr_offset())?;
+        // Read DynamicHistory fields
+        let id_raw = mem_borrow.read_32(hist_addr + probe_id_offset())?;
+        let id = ProbeId::new(id_raw).ok_or_else(|| DebugCollectorError::ProbeIdError)?;
+        let buf_addr = mem_borrow.read_word(hist_addr + log_storage_addr_offset())?;
+        let buf_cap = mem_borrow
+            .read_word(hist_addr + log_storage_cap_offset(hist_addr.size()))?
+            .into();
+        let write_seqn_high_addr = hist_addr + write_seqn_high_offset();
+        let write_seqn_low_addr = hist_addr + write_seqn_low_offset();
+        let overwrite_seqn_high_addr = hist_addr + overwrite_seqn_high_offset();
+        let overwrite_seqn_low_addr = hist_addr + overwrite_seqn_low_offset();
+        let priority_field_addr = hist_addr + overwrite_priority_offset();
+        drop(mem_borrow);
```

Comment:
Rather than using `drop` I recommend instead using a block to nest the borrow and the subsequent actions with it, returning the retrieved values from the block.